qutip/core/data/mul.pyx

```diff
@@ -15,15 +17,24 @@ cpdef CSR imul_csr(CSR matrix, double complex value):
     """Multiply this CSR `matrix` by a complex scalar `value`."""
     cdef idxint l = csr.nnz(matrix)
     cdef int ONE=1
+    if not (isfinite(value.real) and isfinite(value.imag)):
+        raise ValueError(
+            f"Multiplying CSR matrix by non-finite value {value!r}"
+            " would generate a dense matrix of NaNs."
+        )
     zscal(&l, &value, matrix.data, &ONE)
     return matrix
 
 cpdef CSR mul_csr(CSR matrix, double complex value):
     """Multiply this CSR `matrix` by a complex scalar `value`."""
-    if value == 0:
-        return csr.zeros(matrix.shape[0], matrix.shape[1])
```

Comment:
I hate the removal of this shortcut.
`mul(data, 0)` can happen often in QobjEvo when the coefficient is a pulse or step function.
Without this shortcut, they need to loop over the values twice, once for this `mul`, once for the `add` or `matmul` following.
Thus this shortcut will actually help users.
It is also one of the only places where there is a `mul(data, 0)` in the code.

In some theoretical case, it can shallow a NaN, but I don't see any realistic case where this could hide an error.

So everyone loses for a case that will probably never happen...

Can you write a code example where this would save us? Something that looks like it should work, but cause infinities or NaN that would be erased by a multiplication by 0 without returning garbage.

If we can't even think of such a case, let's keep the shortcut.
If you can find one, let's loop over the values to find Nans and return the zeros crs if none are found so we at least loop only once over the values.

Reply:
Maybe we just need to make `_data.zeros[type(x)](*x.shape)` easier to use so that we don't need to write `mul(data, 0)`? I can add `_data.zeros_like(x)` if that would be better?

In the specific case where `mul(data, 0)` is intended to create a matrix of zeros of the right shape, the current fast path can never be wrong since that is exactly what it does.  In general, `mul(data, 0)` is currently just incorrect floating point arithmetic.

I was wondering about adding `has_non_finite` flag to CSRs. It will be a bit of pain to keep track of everywhere but it would allow us to fast path the all finite cases without sacrificing correctness.

If this is a blocker for this PR, I'm happy to undo the current change and just make the tests pass by just explicitly stating that the expected result is `0` in those cases in the test.

I also don't want to sacrifice performance, but it is important to have a clear set of tests that show how we expect the data layer to behave in these edge cases.

Reply:
I am more thinking about the use in `QobjEvo`. I don't think we ever use it as `_data.zeros_like`, in the code base, but that function would be useful, as would be an `identity_like` as we discussed somewhere.

I would be fine with a `has_non_finite` flag, but I think we should raise an error anytime we would raise that flag.
I think that we could do the check when we tidyup: we already check the scale of all values and it's done in most operation that could create zeros, thus functions at risk.

I see a CSR matrix full of zeros as wrong, if not an error, so this is a blocker to me. 
This and tests should pass reliably.

Reply:
I am using `mul` as `_data.zeros_like` in stochastic...